### PR TITLE
rgw: Adding tests for Bucket Policy.

### PR DESCRIPTION
**Pull request by pritha-srivastava**
Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

### CODE REVIEW

File: s3tests/functional/test_s3.py [hunk 1]
```diff
@@ -8266,3 +8266,114 @@ def test_sse_kms_barb_transfer_13b():
...
+@attr(method='get')
+@attr(operation='Test Bucket Policy')
+@attr(assertion='succeeds')
+@attr('bucket-policy')
+def test_bucket_policy():
+    bucket = get_new_bucket(name='my-bucket')
+    key = bucket.new_key('asdf')
+    key.set_contents_from_string('asdf')
+    l = bucket.list()
+    policy_document =	\
```
Comment (pritha-srivastava) [inline s3tests/functional/test_s3.py]: Thank you for the useful suggestion.

File: s3tests/functional/test_s3.py [hunk 2]
```diff
@@ -8266,3 +8266,114 @@ def test_sse_kms_barb_transfer_13b():
...
+        host=s3['alt'].host,
+        calling_format=s3['alt'].calling_format,
+        )
+    e = assert_raises(boto.exception.S3ResponseError, new_conn.get_bucket, 'my-bucket-2')
+    eq(e.status, 403)
+    eq(e.reason, 'Forbidden')
+    eq(e.error_code, 'AccessDenied')
+
+@attr(resource='bucket')
+@attr(method='get')
```
Comment (pritha-srivastava) [inline s3tests/functional/test_s3.py]: Actually, the intention is to check the side-effect of setting a policy on the Bucket List operation, hence I think 'get' will be appropriate to describe this.

Comment (theanalyst) [inline s3tests/functional/test_s3.py]: ack

File: s3tests/functional/test_s3.py (hunk 1 above)
Comment (theanalyst) [inline s3tests/functional/test_s3.py]: also if you use `'''` single quotes you dont have to escape the the double quotes

**Comment by mdw-at-linuxbox** [comment]
I tried this branch of s3-tests on a fairly recent vintage master (bf72bd8d + changes from PR#14997).  My setup happened to have pre-existing data + keystone, so it may have interacted badly with these tests.  It's also possible my setup is simply out of date w/r/t the latest s3-tests.

Anyways, here's what failed for me:
s3tests.functional.test_s3.test_list_buckets_bad_auth 
s3tests.functional.test_s3.test_lifecycle_expiration 
s3tests.functional.test_s3.test_encrypted_transfer_1b 
s3tests.functional.test_s3.test_encrypted_transfer_1kb 
s3tests.functional.test_s3.test_encrypted_transfer_1MB 
s3tests.functional.test_s3.test_encrypted_transfer_13b 
s3tests.functional.test_s3.test_encryption_sse_c_method_head 
s3tests.functional.test_s3.test_encryption_sse_c_present 
s3tests.functional.test_s3.test_encryption_sse_c_other_key 
s3tests.functional.test_s3.test_encryption_sse_c_multipart_upload 
s3tests.functional.test_s3.test_encryption_sse_c_multipart_bad_download 
s3tests.functional.test_s3.test_encryption_sse_c_post_object_authenticated_request 
s3tests.functional.test_s3.test_sse_kms_transfer_1b
s3tests.functional.test_s3.test_sse_kms_transfer_1kb
s3tests.functional.test_s3.test_sse_kms_transfer_1MB
s3tests.functional.test_s3.test_sse_kms_transfer_13b
s3tests.functional.test_s3.test_sse_kms_method_head
s3tests.functional.test_s3.test_sse_kms_present
s3tests.functional.test_s3.test_sse_kms_other_key
s3tests.functional.test_s3.test_sse_kms_multipart_upload
s3tests.functional.test_s3.test_sse_kms_multipart_invalid_chunks_1
s3tests.functional.test_s3.test_sse_kms_multipart_invalid_chunks_2
s3tests.functional.test_s3.test_sse_kms_post_object_authenticated_request
s3tests.functional.test_s3.test_bucket_policy
s3tests.functional.test_s3.test_bucket_policy_acl
s3tests.functional.test_s3.test_bucket_policy_different_tenant

I can provide more details if people want - or retest w/o keystone if folks think that will improve things.

**Comment by adamemerson** [comment]
Maybe try again without keystone. A lot of those failing tests are
nothing we touched. or maybe compare to current s3-tests master.

**Comment by pritha-srivastava** [comment]
@adamemerson, @mattbenjamin  : I had scheduled a job of my own yesterday to make sure the new bucket policy tests work - http://pulpito.ceph.com/prsrivas-2017-06-05_09:56:29-rgw-wip-rgw-s3tests-conf---basic-smithi/, I see errors related to std::regex in the log file. The parts of 'Policy' code that uses std::regex, throws an exception. I was talking to Matt about this in the standup yesterday. He had earlier told that someone had pointed out that std::regex was not implemented fully in some versions of RHEL and CentOS, and that we will have to replace std::regex with its boost counterpart. So I wanted to ask you whether you plan to take on that task? Or I can volunteer to make those changes. I guess we will want to get those changes in, before Luminous?

**Comment by pritha-srivastava** [comment]
The exact error is something like this: 2017-06-05T11:03:41.295 INFO:tasks.rgw.client.0.smithi001.stdout:terminate called after throwing an instance of 'std::regex_error'
2017-06-05T11:03:41.298 INFO:tasks.rgw.client.0.smithi001.stdout:  what():  regex_error
2017-06-05T11:03:41.303 INFO:tasks.rgw.client.0.smithi001.stdout:*** Caught signal (Aborted) **
2017-06-05T11:03:41.303 INFO:tasks.rgw.client.0.smithi001.stdout: in thread 69e3f700 thread_name:civetweb-worker
2017-06-05T11:03:41.390 INFO:tasks.rgw.client.0.smithi001.stdout: ceph version 12.0.2-1492-g2872918 (2872918b0fb1828e02c66cedf5633357bb20e7df)
2017-06-05T11:03:41.391 INFO:tasks.rgw.client.0.smithi001.stdout: 1: (()+0x1f12af) [0x2f92af]
2017-06-05T11:03:41.391 INFO:tasks.rgw.client.0.smithi001.stdout: 2: (()+0xf370) [0x5c92370]
2017-06-05T11:03:41.391 INFO:tasks.rgw.client.0.smithi001.stdout: 3: (gsignal()+0x37) [0x10f201d7]
2017-06-05T11:03:41.391 INFO:tasks.rgw.client.0.smithi001.stdout: 4: (abort()+0x148) [0x10f218c8]
2017-06-05T11:03:41.391 INFO:tasks.rgw.client.0.smithi001.stdout: 5: (__gnu_cxx::__verbose_terminate_handler()+0x165) [0x1072a9d5]
2017-06-05T11:03:41.391 INFO:tasks.rgw.client.0.smithi001.stdout: 6: (()+0x5e946) [0x10728946]
2017-06-05T11:03:41.391 INFO:tasks.rgw.client.0.smithi001.stdout: 7: (()+0x5e973) [0x10728973]
2017-06-05T11:03:41.391 INFO:tasks.rgw.client.0.smithi001.stdout: 8: (()+0x5eb93) [0x10728b93]
2017-06-05T11:03:41.391 INFO:tasks.rgw.client.0.smithi001.stdout: 9: (std::__throw_regex_error(std::regex_constants::error_type)+0x35) [0x1077df95]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 10: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_bracket_expression()+0xc6) [0x50bd96]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 11: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_atom()+0x5a) [0x50bfaa]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 12: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_term()+0x2a) [0x50c64a]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 13: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_alternative()+0x1e) [0x50c6fe]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 14: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_disjunction()+0x24) [0x50c894]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 15: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_atom()+0x461) [0x50c3b1]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 16: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_term()+0x2a) [0x50c64a]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 17: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_alternative()+0x1e) [0x50c6fe]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 18: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_alternative()+0x7f) [0x50c75f]
2017-06-05T11:03:41.392 INFO:tasks.rgw.client.0.smithi001.stdout: 19: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_alternative()+0x7f) [0x50c75f]
2017-06-05T11:03:41.393 INFO:tasks.rgw.client.0.smithi001.stdout: 20: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_alternative()+0x7f) [0x50c75f]
2017-06-05T11:03:41.393 INFO:tasks.rgw.client.0.smithi001.stdout: 21: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_alternative()+0x7f) [0x50c75f]
2017-06-05T11:03:41.393 INFO:tasks.rgw.client.0.smithi001.stdout: 22: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_M_disjunction()+0x24) [0x50c894]
2017-06-05T11:03:41.393 INFO:tasks.rgw.client.0.smithi001.stdout: 23: (std::__detail::_Compiler<char const*, std::regex_traits<char> >::_Compiler(char const* const&, char const* const&, std::regex_traits<char>&, unsigned int)+0x214) [0x50cbe4]
2017-06-05T11:03:41.393 INFO:tasks.rgw.client.0.smithi001.stdout: 24: (std::shared_ptr<std::__detail::_Automaton> std::__detail::__compile<char const*, std::regex_traits<char> >(char const* const&, char const* const&, std::regex_traits<char>&, unsigned int)+0x3b) [0x50cf4b]
2017-06-05T11:03:41.393 INFO:tasks.rgw.client.0.smithi001.stdout: 25: (std::basic_regex<char, std::regex_traits<char> >::basic_regex(char const*, unsigned int)+0x61) [0x50d411]
2017-06-05T11:03:41.393 INFO:tasks.rgw.client.0.smithi001.stdout: 26: (rgw::IAM::ARN::parse(std::string const&, bool)+0x2c4) [0x500874]
2017-06-05T11:03:41.393 INFO:tasks.rgw.client.0.smithi001.stdout: 27: (rgw::IAM::ParseState::do_string(CephContext*, char const*, unsigned long)+0x3a7) [0x501f07]
2017-06-05T11:03:41.394 INFO:tasks.rgw.client.0.smithi001.stdout: 28: (void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseString<96u, rapidjson::GenericStringStream<rapidjson::UTF8<char> >, rgw::IAM::PolicyParser>(rapidjson::GenericStringStream<rapidjson::UTF8<char> >&, rgw::IAM::PolicyParser&, bool)+0x239) [0x50d689]
2017-06-05T11:03:41.394 INFO:tasks.rgw.client.0.smithi001.stdout: 29: (void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseArray<96u, rapidjson::GenericStringStream<rapidjson::UTF8<char> >, rgw::IAM::PolicyParser>(rapidjson::GenericStringStream<rapidjson::UTF8<char> >&, rgw::IAM::PolicyParser&)+0xf6) [0x50def6]
2017-06-05T11:03:41.394 INFO:tasks.rgw.client.0.smithi001.stdout: 30: (void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseObject<96u, rapidjson::GenericStringStream<rapidjson::UTF8<char> >, rgw::IAM::PolicyParser>(rapidjson::GenericStringStream<rapidjson::UTF8<char> >&, rgw::IAM::PolicyParser&)+0x144) [0x50dc14]
2017-06-05T11:03:41.394 INFO:tasks.rgw.client.0.smithi001.stdout: 31: (void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseArray<96u, rapidjson::GenericStringStream<rapidjson::UTF8<char> >, rgw::IAM::PolicyParser>(rapidjson::GenericStringStream<rapidjson::UTF8<char> >&, rgw::IAM::PolicyParser&)+0xf6) [0x50def6]
2017-06-05T11:03:41.394 INFO:tasks.rgw.client.0.smithi001.stdout: 32: (void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseObject<96u, rapidjson::GenericStringStream<rapidjson::UTF8<char> >, rgw::IAM::PolicyParser>(rapidjson::GenericStringStream<rapidjson::UTF8<char> >&, rgw::IAM::PolicyParser&)+0x144) [0x50dc14]
2017-06-05T11:03:41.394 INFO:tasks.rgw.client.0.smithi001.stdout: 33: (rgw::IAM::Policy::Policy(CephContext*, std::string const&, ceph::buffer::list const&)+0x11b) [0x50307b]
2017-06-05T11:03:41.394 INFO:tasks.rgw.client.0.smithi001.stdout: 34: (RGWPutBucketPolicy::execute()+0x71) [0x3a76d1]
2017-06-05T11:03:41.394 INFO:tasks.rgw.client.0.smithi001.stdout: 35: (rgw_process_authenticated(RGWHandler_REST*, RGWOp*&, RGWRequest*, req_state*, bool)+0x162) [0x3cba22]
2017-06-05T11:03:41.394 INFO:tasks.rgw.client.0.smithi001.stdout: 36: (process_request(RGWRados*, RGWREST*, RGWRequest*, std::string const&, rgw::auth::StrategyRegistry const&, RGWRestfulIO*, OpsLogSocket*)+0xb66) [0x3cc7b6]
2017-06-05T11:03:41.395 INFO:tasks.rgw.client.0.smithi001.stdout: 37: (RGWCivetWebFrontend::process(mg_connection*)+0x31c) [0x2b08cc]

**Comment by pritha-srivastava** [comment]
All tests upto the bucket policy tests have passed.

**Comment by mdw-at-linuxbox** [comment]
I'm testing with fc25, so I don't think std::regex is an issue for me.

 I looked through ceph-master/qa-suite and figured out that the tests now require this in ceph.conf:
       rgw crypt s3 kms encryption keys = testkey-1=YmluCmJvb3N0CmJvb3N0LWJ1aWxkCmNlcGguY29uZgo= testkey-2=aWIKTWFrZWZpbGUKbWFuCm91dApzcmMKVGVzdGluZwo=
        rgw crypt require ssl = false

With that string, most of the tests worked for me.  These still fail:
s3tests.functional.test_s3.test_list_buckets_bad_auth ... FAIL
s3tests.functional.test_s3.test_lifecycle_expiration ... FAIL
s3tests.functional.test_s3.test_bucket_policy ... ERROR
s3tests.functional.test_s3.test_bucket_policy_acl ... ERROR
s3tests.functional.test_s3.test_bucket_policy_different_tenant ... ERROR

I think that s3tests.functional.test_s3.test_list_buckets_bad_auth is probably failing because I have keystone.  Error I get is: AssertionError: u'InvalidAccessKeyId' != 'SignatureDoesNotMatch'.  In any case, I assume this has nothing to do with bucket policy.

s3tests.functional.test_s3.test_lifecycle_expiration is probably failing because there's another rgw setting to artificially shorten lifecycle_expiration, which I don't normally enable.  And this is definitely also not related to bucket policy.

test_bucket_policy - fails setting "asdf" key (403 Forbidden) before it gets around to setting any policy.
It's using the hard-wired bucketname "my-bucket" which seems very odd.  Also, it doesn't seem to be getting rid of that bucket after the test run.  I think the tests should be letting get_new_bucket() select a random bucket name (seems to be what most of the other tests do).

test_bucket_policy_acl - same problem except "my-bucket-2".

test_bucket_policy_different_tenant - same problem except "my-bucket-3".

If I remove 'name='my-bucket-X' from these tests two of them still error out, but much further down.  test_bucket_policy_acl() actually succeeds.  test_bucket_policy() fails w/ "403 Forbidden".  Possibly keystone related.  test_bucket_policy_different_tenant fails with "KeyError: 'tenant'".  I'm guessing that wants a configuration value for "tenant", -- what should that be set to?  Is there a "radosgw-admin" command that goes with that?

File: s3tests/functional/test_s3.py [hunk 3]
```diff
@@ -8266,3 +8266,114 @@ def test_sse_kms_barb_transfer_13b():
...
         raise SkipTest
     _test_sse_kms_customer_write(13, key_id = config['main']['kms_keyid'])
+
+@attr(resource='bucket')
+@attr(method='get')
+@attr(operation='Test Bucket Policy')
+@attr(assertion='succeeds')
+@attr('bucket-policy')
+def test_bucket_policy():
+    bucket = get_new_bucket(name='my-bucket')
```
Comment (mdw-at-linuxbox) [inline s3tests/functional/test_s3.py]: I believe this should not specify "name='my-bucket'"

File: s3tests/functional/test_s3.py [hunk 4]
```diff
@@ -8266,3 +8266,114 @@ def test_sse_kms_barb_transfer_13b():
...
+    b = new_conn.get_bucket('my-bucket')
+    b.get_all_keys()
+
+@attr(resource='bucket')
+@attr(method='get')
+@attr(operation='Test Bucket Policy and ACL')
+@attr(assertion='fails')
+@attr('bucket-policy')
+def test_bucket_policy_acl():
+    bucket = get_new_bucket(name='my-bucket-2')
```
Comment (mdw-at-linuxbox) [inline s3tests/functional/test_s3.py]: I believe this should not specify "name='my-bucket-2'"

File: s3tests/functional/test_s3.py [hunk 5]
```diff
@@ -8266,3 +8266,114 @@ def test_sse_kms_barb_transfer_13b():
...
+    eq(e.reason, 'Forbidden')
+    eq(e.error_code, 'AccessDenied')
+
+@attr(resource='bucket')
+@attr(method='get')
+@attr(operation='Test Bucket Policy for a user belonging to a different tenant')
+@attr(assertion='succeeds')
+@attr('bucket-policy')
+def test_bucket_policy_different_tenant():
+    bucket = get_new_bucket(name='my-bucket-3')
```
Comment (mdw-at-linuxbox) [inline s3tests/functional/test_s3.py]: I believe this should not specify "name='my-bucket-3'"

File: s3tests/functional/test_s3.py (hunk 3 above)
Comment (pritha-srivastava) [inline s3tests/functional/test_s3.py]: Well, the functional tests give a provision to create a bucket of a specific name, and I need to use the name further down in the test code, hence I have used it.

**Comment by pritha-srivastava** [comment]
@mdw-at-linuxbox : A tenanted user is needed to be created for the last bucket policy test that needs to go in the s3 test conf file, something along these lines:

[s3 tenant]
## a tenanted user account, used for bucket policy testing
user_id = testx$9876543210abcdef0123456789abcdef0123456789abcdef0123456789abcdef
display_name = tenanteduser
email = tenanteduser@example.com
access_key = HIJKLMNOPQRSTUVWXYZA
secret_key = opqrstuvwxyzabcdefghijklmnopqrstuvwxyzab

**Comment by adamemerson** [comment]
I have a PR for it I'll get in shortly.

**Comment by pritha-srivastava** [comment]
@adamemerson : Thanks, Please let me know, so that I can re-schedule the teuthology run.

**Comment by mdw-at-linuxbox** [comment]
You can find out the name of a bucket by using "bucket.name".  Or, you can call "get_new_bucket_name()" directly to get the "next bucket name", save it and use that - that's what several other tests do.

There's logic in s3tests/functional/__init__.py to nuke test buckets.  If you create buckets that match the prefix they'll be cleaned up.  If you create buckets with random names they won't be cleaned up.  Buckets that are left around will probably cause the test to fail on repeated runs.  Like the runs I was just doing to test to see if your tests worked.  Or imagine at some point in the future we might want to run s3-tests over and over and over -- say, as part of some kind of stress testing.

**Comment by mdw-at-linuxbox** [comment]
I've run unmodified tests twice on a freshly initialized radosgw setup w/ no keystone.  First time, only test_lifecycle_expiration failed.  2nd time, test_bucket_policy and test_bucket_policy_different_tenant come up as errors.  "radosgw-admin bucket list" clearly shows the 4 buckets "my-bucket", "my-new-bucket", "my-bucket-2" and "my-bucket-3" as left existing after the test is run.

**Comment by pritha-srivastava** [comment]
@mdw-at-linuxbox : I have made the neccessary changes.

File: s3tests/functional/test_s3.py [hunk 6]
```diff
@@ -8266,3 +8266,166 @@ def test_sse_kms_barb_transfer_13b():
...
+            "arn:aws:s3:::*",
+            "arn:aws:s3:::*/*"
+          ]
+        }]
+     })
+    bucket1.set_policy(policy_document)
+
+    json_policy = bucket1.get_policy()
+    bucket2.set_policy(json_policy)
+
```
Comment (adamemerson) [inline s3tests/functional/test_s3.py]: This is excellent and exactly what I was hoping for, thank you.

**Pull request merged**